common/remote/hooks/azure/azurefile.cpp

```diff
@@ -601,112 +596,170 @@ AzureFile::AzureFile(const char *_azureFileName) : fullName(_azureFileName)
+std::string AzureFile::getBlobUrl()
+{
+    if (blobUrl.size()==0)
+        blobUrl = ::getBlobUrl(accountName, containerName, blobName);
+    return blobUrl;
+}
 
-        //MORE: Need to update the code / code_names so the following works:
-        if (streq(ret.error().code.c_str(), "ContainerAlreadyExists"))
-            return true;
-        if (!streq(ret.error().code.c_str(), "ContainerBeingDeleted"))
-            return true;
-        MilliSleep(100);
-    }
+std::shared_ptr<BlobContainerClient> AzureFile::getBlobContainerClient()
+{
+    auto cred = getCredentials();
+    std::string blobContainerUrl = getContainerUrl(accountName, containerName);
+    return std::make_shared<BlobContainerClient>(blobContainerUrl, cred);
 }
 
+template<typename T>
+std::shared_ptr<T> AzureFile::getClient()
+{
+    auto cred = getCredentials();
+    return std::make_shared<T>(getBlobUrl(), cred);
+}
+
+bool AzureFile::createDirectory()
+{
+    auto blobContainerClient = getBlobContainerClient();
+    try
+    {
+        Azure::Response<Models::CreateBlobContainerResult> result = blobContainerClient->CreateIfNotExists();
+        if (result.Value.Created==false)
+            OERRLOG("Azure create container: container not created");
+        return result.Value.Created;
+    }
+    catch (const Azure::Core::RequestFailedException& e)
+    {
+        IException * error = makeStringExceptionV(1234, "Azure create container failed: %s (%d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+        throw error;
+    }
+}
 
 void AzureFile::createAppendBlob()
 {
-    auto client = getClient();
-    auto ret = client->create_append_blob(containerName, blobName).get();
-    checkError(ret);
+    auto appendBlobClient = getClient<AppendBlobClient>();
+    try
+    {
+        Azure::Response<Models::CreateAppendBlobResult> result = appendBlobClient->CreateIfNotExists();
+        if (result.Value.Created==false)
+            OERRLOG("Azure append blob (container %s blob %s): blob not created", containerName.str(), blobName.str());
+        else
+        {
+            setProperties(0, result.Value.LastModified, result.Value.LastModified);
+        }
+    }
+    catch (const Azure::Core::RequestFailedException& e)
+    {
+        IException * error = makeStringExceptionV(1234, "Azure create append blob failed: %s (%d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+        throw error;
+    }
 }
 
 
 void AzureFile::appendToAppendBlob(size32_t len, const void * data)
 {
-    auto client = getClient();
-    std::istringstream input(std::string((const char *)data, len));
-    //implement append_block_from_buffer based on upload_block_from_buffer
-    auto ret = client->append_block_from_stream(containerName, blobName, input).get();
-    checkError(ret);
+    auto appendBlobClient = getClient<AppendBlobClient>();
+    try
+    {
+        Azure::Core::IO::MemoryBodyStream content(reinterpret_cast <const uint8_t *>(data), len);
```

Comment:
Does this clone the data?  If so ddd a comment.  In the future, this would be better if it passed a class instance derived from Azure::Core::IO::BodyStream that did not clone the data